docs/sdg/sdg-flow-yaml.md

```diff
@@ -0,0 +1,101 @@
+
+## Objective
+
+- Library users can specify a custom flow using a well-defined file format.
+- Library users can either use a custom flow standalone, or combine a custom flow with existing flows.
+- The file format and library can evolve substantially without making breaking changes.
+- Incompatible changes can be introduced while retaining support for existing custom flows for a deprecation period.
+
+## Proposal
+
+We will add an API that instantiates a `Flow` object from a file:
+
+```python
+flow = Flow.from_file('mycustomflow.yaml', params)
+block_configs = flow.render()
+```
+
+The YAML file format will mirror the API and look like this:
+
+```yaml
+block_configs:
+  - block_type: LLMBlock
+    block_config: # LLMBlock constructor kwargs
+      block_name: gen_knowledge
+      output_cols: ["question", "response"]
+    gen_kwargs: # kwargs for block.generate()
+      max_tokens": 2048,
+    drop_duplicates: ["question"]
+  - block_type: FilterByValueBlock
+    block_config:
+      block_name: filter_faithfulness
+      filter_column": judgment
+      filter_value: YES
+      operation: operator.eq
+      batch_kwargs:
+        num_procs: 8
+     drop_columns: ["judgment", "explanation"]
+```
+
+### Flow Params
+
+The following parameters are supplied to a Flow constructor and used by the render() method:
+
+- client - an OpenAI completions API client for talking to the teacher model via the serving backend (i.e.  llama-cpp or vLLM)
+- model_family - e.g. mixtral or merlinite
+- model_id - a path name for the specific teacher model being used
+- num_iters - how many samples to generate
+- batched - whether the model serving backend supports batched mode
+
+TODO: figure out how these should be supplied to the blocks, and whether any YAML templating based on these are necessary.
+
+### Model Prompts
+
+Based on whether model_family is mixtral or merlinite, a different prompt is used with the teacher model
+
+```python
+_MODEL_PROMPT_MIXTRAL = "<s> [INST] {prompt} [/INST]"
+_MODEL_PROMPT_MERLINITE = "'<|system|>\nYou are an AI language model developed by IBM Research. You are a cautious assistant. You carefully follow instructions. You are helpful and harmless and you follow ethical guidelines and promote positive behavior.\n<|user|>\n{prompt}\n<|assistant|>\n'"
+```
+
+TODO: figure out how to include the appropriate prompt in the block config.
+
+### Prompt Config Files
+
+Every LLMBlock references a separate prompt config file, and presumably a custom pipeline will provide custom prompt configs too.
+
+These prompt config files are quite simple YAML files - they contain a single object with system, introduction, principles, examples, and generation keys. See e.g. src/instructlab/sdg/configs/skills/freeform_questions.yaml
+
+TODO: figure out how we want to handle these.
```

Comment:
The ones included in the library get included in the Python package and installed alongside the python source. There's a little bit of magic in the code to make sure we load it from the right place from the library's installed location.

For a custom flow, I think a path to a yaml file is still good. Where the files are installed, how they get there ... out of scope for the purposes of the CLI and sdg library, I think.

Reply:
Resolved with:

> We will continue to use these config files unchanged, and custom files can be specified with an absolute path. Relative paths are assumed to be relative to the Python package e.g. `configs/skills/...`.